model_compression_toolkit/core/pytorch/pytorch_device_config.py

```diff
@@ -0,0 +1,118 @@
+
+from model_compression_toolkit import logger
+from model_compression_toolkit.core.pytorch.constants import CUDA, CPU
+
+
+class DeviceManager:
+    """
+    A singleton class to manage the PyTorch device (CPU or CUDA) across different parts of the project.
+    Ensures that only one instance of this class is created and used throughout all files.
+    """
+
+    _instance = None
+
+    def __new__(cls):
+        """
+        Override the __new__ method to implement the singleton pattern.
+        Ensures that only one instance of DeviceManager is created.
+        """
+        if cls._instance is None:
+            cls._instance = super(DeviceManager, cls).__new__(cls)
+            # Initialize the default device
+            cls._instance.DEVICE = torch.device(CUDA if torch.cuda.is_available() else CPU)
+        return cls._instance
+
+    def set_device(self, device_name: str):
+        """
+        Set the device for PyTorch operations.
+
+        Args:
+            device_name (str): The name of the device, e.g., 'cuda:0' or 'cpu'.
+
+        If the specified device is not valid or available, it prints an error message without changing the current device.
+        """
+        is_valid, message = self.is_valid_device(device_name)
+        if is_valid:
+            self.DEVICE = torch.device(device_name)
+        else:
+            logger.Logger.warning(message)
+
+    def get_device(self):
+        """
+        Get the current PyTorch device.
+
+        Returns:
+            torch.device: The current device set for PyTorch operations.
+        """
+        return self.DEVICE
+
+    @staticmethod
+    def is_valid_device(device_name: str):
+        """
+        Check if the specified device name is valid and available.
+
+        Args:
+            device_name (str): The name of the device to check, e.g., 'cuda:0' or 'cuda'.
+
+        Returns:
+            tuple: (bool, str) A tuple where the first element is a boolean indicating if the device is valid,
+                   and the second element is a message describing the validity or the issue.
+        """
+        if CUDA in device_name:
+            if not torch.cuda.is_available():
+                return False, "CUDA is not available"
+
+            if device_name == CUDA:
+                # 'cuda' without a specific index is valid if CUDA is available
+                return True, "Valid device"
+
+            try:
+                device_index = int(device_name.split(':')[1])
+                if device_index >= torch.cuda.device_count():
+                    return False, f"CUDA device index {device_index} out of range. Number of valid devices: {torch.cuda.device_count()}"
+            except IndexError:
+                # Handle cases where the device name is incorrectly formatted
+                return False, "Invalid CUDA device format. Use 'cuda' or 'cuda:x' where x is the device index."
+
+            return True, "Valid device"
+
+        return True, "Valid device"
```

Comment:
What if it's 'cpu'? I think it should be checked here before True is returned

Reply:
There isn't a check for CPU because it is always a valid choice. This checks that if you choose a GPU it exists.

Reply:
I think it's better to verify that it's 'cpu' and not unrelated string, but it's up to you